feature: Optimize Views for Landscape

wba2hi opened the pull request:
Interesting lectures and documentation: 
- https://www.youtube.com/watch?v=m7OL-mVh1E8
- https://www.youtube.com/watch?v=fQjDtCtri4s
- https://m3.material.io/foundations/layout/understanding-layout/overview

What did I use:
- WindowSizeClass to differentiate between Layout and Portrait mode
- FlowRow and FlowColumn Views for "better flow" on changing screens

What did I not use:
- A Canonical Layout. Feed does not fit, List-Detail also does not fit. Supporting Pane could fit, however the RamsesView can only "shrink" in size, it can not "increase" in size 100\*100 -> 100\*50 works but 100\*50 -> 100\*100 (due to expanding shrinking does not work). Therefore we need something to overlay our view and not to reduce it's size. An element which can expand and shrink again. So I added a custom BottomSheet
-> I added AdaptiveViews, which based on the windowSizeClasses show different content (e.g. horizontal orientation instead of vertical), e.g. an AdaptiveNavigation consists of a HorizontalNavigation and a VerticalNavigation which both take there data from the same dataset. 

Why AdaptiveViews? Why 2 Views and not if-else?
I tried to keep everything in one view, see AdaptiveSheet. However when switching from Portrait to Landscape a lot of parameter change. When setting the width in Portrait, we need to set the Height in Landscape for example. However we can not easily use kotlins .apply {} or .also {} scope when working with the modifier. It adds a lot of complexity (if portrait use width, else use height) to each view if it tries to do both. I tried to find the best way by adding two separate views, which are "simpler" and "more maintainable" over one view which tries to do both but is error-prone hard to read and less maintainable. 

I tried to move all "AdaptiveView"-logic inside the AdaptiveScreen
I moved the settings from being a FloatingActionButton inside the RamsesView to an element inside the NavigationBar

Closes: #20

wba2hi committed:
feature: Optimize Views for Landscape

Closes: #20
Signed-off-by: Andre Weber <andre.weber3@etas.com>

wba2hi committed:
chore: Fix Lint Issues

wba2hi committed:
chore: Move Library into .toml

wba2hi committed:
chore: Fix z-indexes of Views

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/extension/ConfigurationExtension.kt:
```diff
@@ -0,0 +1,36 @@
+import androidx.compose.ui.unit.dp
+
+@OptIn(ExperimentalMaterial3WindowSizeClassApi::class)
+@Composable
```
Can you explain why is this a composable extension? Does a val extension "windowSizeClass" work here too? Having getter extension feels not so kotlin like.

```
@OptIn(ExperimentalMaterial3WindowSizeClassApi::class)
val Configuration.getWindowSizeClass: WindowSizeClass
    get() {
        val screenWidth = screenWidthDp
        val screenHeight = screenHeightDp
        val size = DpSize(screenWidth.dp, screenHeight.dp)
        return WindowSizeClass.calculateFromSize(size)
    }
```

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/extension/ConstraintScopeExtension.kt:
```diff
@@ -0,0 +1,93 @@
+import org.eclipse.kuksa.companion.feature.door.view.verticalMarginAnchorToBackDoor
+import org.eclipse.kuksa.companion.feature.door.view.verticalMarginAnchorToDoor
+
+fun ConstrainScope.alignDriverFrontDoor(
```
I am thinking if there is a better call than an extension file here. Could we have a class like "DoorLockOverlayWindow" which takes the WindowSizeClass + AnchorPoint all overlay icons  which defines the "horizontalMarginAnchorToDoor" + the vertical + horizontal anchor? The following code would then sit in the compose function 

`end.linkTo(doorLockOverlay.driverFrontDoor.horizontalAnchor, doorLockOverlay.driverFrontDoor.horizontalMargin)`

The extension file could be removed then and in the future we just have to define a new class for new overlay constraints?

wba2hi replied:
This class was the last one I created before putting the ticket to review. Reason should be clear. Remove redundancy and make the "put some element at a specific place (e.g. next to the DriverFrontDoor) without having to copy paste code" as easy as possible. 
 
I probably don't understand the overall picture of your snippet yet, but it looks like it re-introduces redundancy again because each Overlay class needs to manually add their margins / anchorPoints again. I also don't understand the role of the DoorLockOverlayWindow yet. However I am really concerned about adding ANOTHER layer on top of the Overlay.

Please elaborate or give a more in-depth sample because I agree that - while the chosen solution can easily be changed - it's not optimal because it's much to specific and just clutters the ConstrainScope.

Chrylo replied:
I imagined something like this:

It is just refactoring an extension into a class. My main gripe with this extension is that it feels like we should link the RamsesElements with the constraints and not the other way around. It should also be obvious to the devs and easily expandable if we want to add more anchor points of the car (Where extensions feel more hidden here than classes). 

Currently the extension is also available for every compose view with a constraint (which may have nothing to do with Ramses). To have a function like "alignDriverFrontDoor" feels off. 

Maybe you have some ideas to improve this idea even further if you agree. :)

```
// Compose class
val ramsesAnchor = DriverFrontDoorRamsesAnchor(windowSizeClass, anchorPoint)
ramsesAnchor.align(this)

abstract class RamsesAnchor(
    protected val windowSizeClass: WindowSizeClass,
    protected val anchorPoint: ConstrainedLayoutReference,
) {
    fun align(constrainScope: ConstrainScope) {
        when (windowSizeClass.widthSizeClass) {
            WindowWidthSizeClass.Compact -> alignCompact(constrainScope)
            else -> alignMedium(constrainScope)
        }
    }

    protected abstract fun alignCompact(constrainScope: ConstrainScope)

    protected abstract fun alignMedium(constrainScope: ConstrainScope)
}

class DriverFrontDoorRamsesAnchor(windowSizeClass: WindowSizeClass, anchorPoint: ConstrainedLayoutReference) :
    RamsesAnchor(windowSizeClass, anchorPoint) {

    override fun alignCompact(constrainScope: ConstrainScope) {
        constrainScope.apply {
            end.linkTo(anchorPoint.start, horizontalMarginAnchorToDoor)
            top.linkTo(anchorPoint.bottom, verticalMarginAnchorToDoor)
        }
    }

    override fun alignMedium(constrainScope: ConstrainScope) {
        constrainScope.apply {
            end.linkTo(anchorPoint.start, verticalMarginAnchorToDoor)
            bottom.linkTo(anchorPoint.top, horizontalMarginAnchorToDoor)
        }
    }
}
```

wba2hi replied:
I like this idea (and I implemented it as suggested). Only thing I don't like is that eventhough the number of instances are basically fixed, we have to create a new object each time.

```
val driverFrontDoorAnchor = DriverFrontDoorAnchor(windowSizeClass, anchorPoint)   
driverFrontDoorAnchor.align(this)
```

Only way to fix this would be by using enums and putting all parameters into the "align" method.

```
enum class RamsesAnchor {
   DRIVER_FRONT_DOOR, // implement alignCompact and alignMedium
   PASSENGER_FRONT_DOOR,
   DRIVER_BACK_DOOR
   PASSENGER_BACK_DOOR,
   TRUNK,

   fun align(windowSizeClass: WindowSizeClass, anchorPoint: ConstrainedLayoutReference, constrainScope: ConstrainScope) {
...
... rest as before
...
    }
}
```

However I think the signature which you showed feels more natural and sane and therefore left it as is

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/feature/connection/view/AnimatedLoadingText.kt:
```diff
@@ -0,0 +1,58 @@
+import org.eclipse.kuksa.companion.feature.connection.viewModel.ConnectionStatusViewModel.ConnectionState
+import kotlin.time.Duration.Companion.milliseconds
+
+private const val MAX_NUMBER_OF_DOTS = 3
```
Composable const are always camelcase

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/feature/connection/view/AnimatedLoadingText.kt:
```diff
@@ -0,0 +1,58 @@
+
+@Composable
+fun animateLoadingText(
+    connectionState: ConnectionState,
```
Can we uncouple the connection state here? It is an "AnimatedLoadingText". It is either animating or not. Maybe just a "isAnimating" flag is enough. I think the logic for converting the ConnectionState to the bool should be in the actual ConnectionView.

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/feature/connection/view/VerticalConnectionStatusView.kt:
```diff
@@ -0,0 +1,125 @@
+import org.eclipse.kuksa.companion.feature.connection.viewModel.ConnectionStatusViewModel
+
+@Composable
+fun VerticalConnectionStatusView(
```
What do you think about introducing a sub package for landscape and portrait in the view section? I think it makes the pattern clearer to use separate compose files for these cases.

wba2hi replied:
In general a good idea, however imo not applicable for our use case the since we already splitted everything up by feature. We have 7 features, and by splitting this stuff up again by landscape portrait we have basically 14 packages (landscape, portrait) in which a single view is put plus we have to traverse the folder structure through these 14 folders when navigating from file to file. 

It would have worked to split up by orientation instead of by feature

Therefore I prefer to keep the stuff together and not introduce packages for single files. 

> [...] makes the pattern clearer to use separate compose files for these cases

I don't see any clarity added by a landscape / portrait package when the Horizontal- / Vertical- / Adaptive- Views can reside next to each other to be honest.

Chrylo replied:
OK for me

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/feature/connection/view/VerticalConnectionStatusView.kt:
```diff
@@ -0,0 +1,125 @@
+
+@Preview(heightDp = 300)
+@Composable
+private fun VerticalConnectionStatusPreview_Disconnected() {
```
Can we align the naming here too like in the other Ticket. No underscore :)?

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/feature/connection/view/VerticalText.kt:
```diff
@@ -0,0 +1,81 @@
+import androidx.compose.ui.unit.TextUnit
+
+@Composable
+fun VerticalText(
```
This does not have to be inside a landscape package. Could even be extracted from the connection package since its also usable somewhere else.

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/feature/connection/view/VerticalText.kt:
```diff
@@ -0,0 +1,81 @@
+    charArray.forEach { verticalTextBuilder.append(it).append(System.lineSeparator()) }
+    val verticalText = verticalTextBuilder.toString()
+
+    Text(
```
The class feels a bit redundant with all the parameters. Can we create a compose method like the "animateText" -> "convertToVertical" Text extension which returns a text.

wba2hi replied:
Created a String Extension instead

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/feature/connection/view/AdaptiveConnectionStatusView.kt:
```diff
@@ -0,0 +1,84 @@
+    windowSizeClass: WindowSizeClass,
+    modifier: Modifier = Modifier,
+) {
+    if (windowSizeClass.widthSizeClass == WindowWidthSizeClass.Compact) {
```
I personally don't think its a neat solution but seems to be the way to go:
https://developer.android.com/jetpack/compose/layouts/adaptive

Extracting the logic to different classes is making this way better at least :)

Chrylo commented:
Regarding the custom BottomSheet. I really liked how the bottom sheet felt. Can we maybe revert to that one and use a Sidesheet like in the link?

<img width="813" alt="Screenshot 2024-02-05 at 13 27 44" src="https://github.com/eclipse-kuksa/kuksa-android-companion/assets/1355532/7da64c5a-a97e-4c7f-bcfd-4550c1ba6701">


https://m2.material.io/components/sheets-bottom#behavior

Another idea would be to use an expanding Action Button like this:
https://stackoverflow.com/questions/75933627/how-to-expand-fab-into-a-sheet-in-jetpack-compose

Chrylo commented:
Functional: 

- I would not expect that a reconnection (unsubscribe / disconnect etc.) happens during a rotation because only the views should be rebuild. Probably not easy to fix if something is a bit too entangled here. Lets discuss this.
- The current tab is not saved when rotating
- The Tab bar has now more height since the settings button was added there. Is it possible to reduce that again? Is it possible to add some kind of line divider between the settings and the previous button? To show some visual decoupling?

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/feature/home/view/sheet/AdaptiveLine.kt:
```diff
@@ -0,0 +1,110 @@
+ * horizontal line.
+ */
+@Composable
+fun AdaptiveLine(
```
Could be put in another generic package outside the "home" feature maybe

wba2hi replied:
I think this class is much to specific and bound to the Sheet to be used outside of the Sheet Context and therefore would keep it there for the time being. 

However, if we change our Sheet Behavior it might be that this class will be removed anyway. So let's wait and see for now.

wba2hi replied:
AdaptiveLine was removed

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/feature/home/view/sheet/AdaptiveSheetLayout.kt:
```diff
@@ -0,0 +1,114 @@
+    ConstraintLayout(modifier = modifier.fillMaxSize()) {
+        val sheetRef = createRef()
+
+        val sheetModifier = when (windowSizeClass.widthSizeClass) {
```
I love the when for the windowSizeClass instead of if else :)

wba2hi committed:
chore: Change ConfigurationExtension from fun to val

wba2hi committed:
chore: Decouple animateLoadingText Logic from ConnectionState

wba2hi committed:
chore: Rename Method Name of Previews

wba2hi committed:
chore: Replace VerticalText with Extension

wba2hi committed:
chore: Save NavigationBar Tab Information

wba2hi committed:
chore: Align Naming

wba2hi committed:
chore: Re-Add BottomSheet

wba2hi committed:
chore: Replace ConstraintScopeExtension with RamsesAnchor

wba2hi committed:
chore: Change Size of NavigationBarIcons

wba2hi committed:
chore: Add SideSheet to Landscape Mode

wba2hi committed:
chore: Fix Lint Findings

wba2hi committed:
chore: Disable KTLint Naming Rule

KTLints property-naming rule is not compatible with compose.
The official recommendation is to disable the rule, see:
https://github.com/pinterest/ktlint/issues/2139

This should be okay, since we still have the detekt linter, which
also checks for names in a more compose compatible way.

wba2hi committed:
chore: Keep Connection Alive on Orientation Change

wba2hi committed:
chore: Remove Legacy Code

wba2hi commented:
> * I would not expect that a reconnection (unsubscribe / disconnect etc.) happens during a rotation because only the views should be rebuild. Probably not easy to fix if something is a bit too entangled here. Lets discuss this.

Fixed

> * The current tab is not saved when rotating

Fixed
 
> * The Tab bar has now more height since the settings button was added there. Is it possible to reduce that again? Is it possible to add some kind of line divider between the settings and the previous button? To show some visual decoupling?

Previously we used a TabHost, now we use a NavigationBar which has a well-defined height.  Therefore the new height is intended. Increased the Icon size a bit to compensate for the height of the NavigationBar.

wba2hi committed:
chore: Remove onSideSheetInitialised Callback

wba2hi committed:
chore: Remove Unused Resources

wba2hi committed:
chore: Change isBottomSheetEnabled to more generic isSheetEnabled

wba2hi committed:
chore: Center NavigationRail

wba2hi committed:
chore: Add Rounded Corner to SideSheet

wba2hi committed:
chore: Change Navigation Labels and add Padding

wba2hi committed:
chore: Correctly Register DisconnectListener on Orientation Change

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/MainActivity.kt:
```diff
@@ -85,7 +91,13 @@ class MainActivity : ComponentActivity() {
 
-    private var dataBrokerConnection: DataBrokerConnection? = null
+    // storing the connection in the Application keeps the Connection alive on orientation changes
+    private var dataBrokerConnection: DataBrokerConnection?
```
In the future we can use Hilt to simplify this and just inject the same instance. But fine for now :)

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/MainActivity.kt:
```diff
@@ -286,6 +301,12 @@ class MainActivity : ComponentActivity() {
     private fun connectToDataBroker(onConnected: () -> Unit = {}) {
+        // dataBrokerConnection is already established e.g. after an orientation change
+        if (dataBrokerConnection != null) {
+            onConnected()
```
Is the onConnected() correct here? Doesn't it lead to duplicate subscriptions?

I am still seeing a lot of "Remove subscription", "Subscribe to" when changing the orientation. I also would expect that this does not happen during orientation change. Probably not easy to fix so it would be fine for me to do it in another ticket in addition to the mentioned Hilt improvements.

wba2hi replied:
Our Listeneners (vssDoorListener, vssTemperatureListener, vssLightListener, ...) are stored inside the MainActivity, which is destroyed due to the orientation change. In onDestroy we cleanup by removing all subscriptions (unsubscribe) again, which is fine from a lifecycle point of view (subscribe onConnected() <-> unsubscribe onDestroy) to prevent leaking of any ressources. 

> Is the onConnected() correct here?

Yes, the new "onConnected()" call in connectToDataBroker is required to re-register the listeners and bring the MainActivity back into a working state. 

> Doesn't it lead to duplicate subscriptions?

No, since we cleanup in onDestroy the old subscriptions. 

To fix this issue we would need to elevate the logic from the MainActivity into the application, or probably better (?) into the viewModels and remove the "unsubscribe" call from MainActivity#onDestroy.

We should test this thoroughly and also check all other properties in MainActivity if they are stateful or not. As you suggested, let's do this in another ticket.

wba2hi replied:
Created #55

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/extension/StringExtension.kt:
```diff
@@ -0,0 +1,32 @@
+
+package org.eclipse.kuksa.companion.extension
+
+fun String.convertToVerticalString(): String {
```
Maybe "toVertical" is good enough as a name? :) Feels short and neat as an extension name

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/feature/home/view/navigation/NavigationViewModel.kt:
```diff
@@ -0,0 +1,31 @@
+ *
+ */
+
+package org.eclipse.kuksa.companion.feature.home.view.navigation
```
What do you think about moving navigation out of the home package and add the viewModel / view / model folder? To be consistent with the package pattern? 

optional: Because it's a small sub package it may stay like this (for now).

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/feature/home/view/sheet/SideSheetView.kt:
```diff
@@ -0,0 +1,179 @@
+/*
```
The preview is broken for me

wba2hi replied:
Android Studio Issue: 
https://stackoverflow.com/a/77484044

Update your Android Studio to Iguana 2023.2.1 RC 1 which supports Rendering Api Versions up to 34

Chrylo commented on app/src/main/kotlin/org/eclipse/kuksa/companion/feature/home/view/sheet/SideSheetView.kt:
```diff
@@ -0,0 +1,179 @@
+            val paddingValues = PaddingValues(0.dp)
+            content(paddingValues)
+        }
+        if (isSideSheetEnabled) {
```
Can we do an return here? I guess the happy path is that it should be enabled. :)

wba2hi committed:
chore: Align Feature Folder Structure

- add feature folder for Navigation and (Side-)/(Bottom-)Sheet

wba2hi committed:
chore: Use Early Return if SideSheet is disabled

wba2hi committed:
chore: Align Subscribe and Unsubscribe

wba2hi committed:
Merge branch 'main' into feature-20

* main:
  feature: Add Icon for F-Droid to Fastlane
  chore: Add Info about KUKSA CLI and Python Library
  chore: Drop ".val" suffix
  docs: Adapt Documentation as per Review Findings
  docs: Add Quickstart Guide
  feature(Build): Remove F-Droid BuildType

Chrylo commented:
LGTM 

- Tested all views in both landscape + portrait.
- Tested connection to the Databroker while rotating the views
- Tested new Sidesheet element
- Tested state restoration while rotating the views

This was a big task so some improvements will be made in follow up tickets like #55.

known issues:
- The settings view may be squeezed for some text boxes depending on the phone
- The app resubscribes to all properties on orientation change: #55
- The temperature view can't be dragged in landscape mode because the event is overridden by the Sidesheet
